packages/cli/src/databases/repositories/workflow.repository.ts

```diff
@@ -99,30 +121,186 @@ export class WorkflowRepository extends Repository<WorkflowEntity> {
+				.addCommonTableExpression(
+					`SELECT * FROM ${qb.escape('FOLDERS_QUERY')} UNION ALL SELECT * FROM ${qb.escape('WORKFLOWS_QUERY')}`,
+					'RESULT_QUERY',
+				),
+			sortByColumn,
+			sortByDirection,
+		};
+	}
+
+	async getWorkflowsAndFoldersUnion(workflowIds: string[], options: ListQuery.Options = {}) {
+		const { baseQuery, sortByColumn, sortByDirection } = this.buildBaseUnionQuery(
+			workflowIds,
+			options,
+		);
+
+		const query = baseQuery
+			.select(`${baseQuery.escape('RESULT')}.*`)
+			.from('RESULT_QUERY', 'RESULT');
+		if (sortByColumn === 'name') {
+			query
+				.addSelect(`LOWER(${baseQuery.escape('RESULT')}.${baseQuery.escape('name')})`, 'name_lower')
+				.orderBy('name_lower', sortByDirection);
+		} else {
+			query.orderBy(
+				`${baseQuery.escape('RESULT')}.${baseQuery.escape(sortByColumn)}`,
+				sortByDirection,
+			);
+		}
+
+		if (options.take) {
+			query.take(options.take);
+		}
+
+		query.skip(options.skip ?? 0);
+
+		const workflowsAndFolders = await query.getRawMany<WorkflowAndFolderUnion>();
+
+		return workflowsAndFolders.map((item) => {
+			const { name_lower, ...rest } = item;
+			return rest;
+		});
+	}
+
+	async getWorkflowsAndFoldersCount(workflowIds: string[], options: ListQuery.Options = {}) {
+		const { skip, take, ...baseQueryParameters } = options;
+
+		const { baseQuery } = this.buildBaseUnionQuery(workflowIds, baseQueryParameters);
+
+		const response = await baseQuery
+			.select(`COUNT(DISTINCT ${baseQuery.escape('RESULT')}.${baseQuery.escape('id')})`, 'count')
+			.from('RESULT_QUERY', 'RESULT')
+			.select('COUNT(*)', 'count')
+			.getRawOne<{ count: number | string }>();
+
+		return Number(response?.count) || 0;
+	}
+
+	async getWorkflowsAndFoldersWithCount(workflowIds: string[], options: ListQuery.Options = {}) {
+		const [workflowsAndFolders, count] = await Promise.all([
+			this.getWorkflowsAndFoldersUnion(workflowIds, options),
+			this.getWorkflowsAndFoldersCount(workflowIds, options),
+		]);
+
+		const { workflows, folders } = await this.fetchExtraData(workflowIds);
+
+		const enrichedWorkflowsAndFolders = this.enrichDataWithExtras(workflowsAndFolders, {
+			workflows,
+			folders,
+		});
+
+		return [enrichedWorkflowsAndFolders, count] as const;
+	}
+
+	private async fetchExtraData(workflowIds: string[]) {
+		const [workflows, folders] = await Promise.all([
+			this.getMany(workflowIds),
+			this.folderRepository.getMany(),
+		]);
+
+		return { workflows, folders };
+	}
+
+	private enrichDataWithExtras(
+		baseData: WorkflowAndFolderUnion[],
+		extraData: {
+			workflows: ListQuery.Workflow.WithSharing[] | ListQuery.Workflow.Plain[];
+			folders: Folder[];
+		},
+	): WorkflowAndFolderUnionFull[] {
+		return baseData.map((item) => {
+			const extraItem =
+				item.resource === 'folder'
+					? extraData.folders.find((folder) => folder.id === item.id)
+					: extraData.workflows.find((workflow) => workflow.id === item.id);
```

Comment:
Can we build a map of these so we are not doing a linear time lookup on every iteration?

Reply:
Good point.